### DISCUSSION TITLE
app_.getInitialProps with page getServersideProps

### DISCUSSION THREAD

**Discussion by majelbstoat** [vercel/next.js#11183] [Help]
I know that `_app.getServersideProps` doesn't exist yet, but is there a way to get `_app.getInitialProps` to correctly pass props to the `App` if a page defines `getServersideProps`? Simple illustration:

```tsx
// _app.tsx
const MyApp = ({ Component, pageProps }: AppProps) => {
  const { viewer } = pageProps
  console.log('viewer', pageProps)

  return (
    <UserProvider viewer={viewer}>
      <Component {...pageProps} />
    </UserProvider>
  )
}

MyApp.getInitialProps = async (appContext: AppContextType<Router>) => {
  const appProps = await App.getInitialProps(appContext)
  appProps.pageProps.viewer = { id: 'monkey' }
  return { ...appProps }
}

// index.tsx
const IndexPage: NextPage = ({viewer}: IndexPageProps) => {
  return (
    <h1>OK</h1>
  )
}
```
This prints `viewer { id: 'monkey' }` as expected. However, if I add getServersideProps to IndexPage:

```tsx
// index.tsx
export const getServerSideProps: GetServerSideProps = async () => {
  return { props: {} }
}
```

It now prints `viewer undefined`

I presume the result of `getServersideProps` is being used as the page props, but I don't understand the order in which things execute, and/or whether something like this is possible. I don't want to use `getInitialProps` on the page, because `getServersideProps` does exactly what I want in terms of direct db access etc. (Nice capability upgrade, thank you!)

What I'm trying to achieve:

* load a user from a cookie on the server side
* use that user to set an initial client side value for a viewer/setViewer state on a context Provider component high up in the react tree.
* add the viewer to the cache key for SWR, so that data is automatically refetched when we call setViewer and the viewer changes (on login/logout without redirect).
* still have all this server side rendered.

**Comment by CurtisBaldwinson** [comment]
You can observe the following results printed in your browser developer console with the following code sample:
```
MyApp.props {
  serverSide: true,
  pageProps: {
    somePageProp: "somePagePropContent"
  }
  ...[other prop data such as Component, err, etc.]
}
```

```
# _app.js
function MyApp(props) {
  console.log("MyApp.props", props);

  return (
    <div>
      <props.Component {...props.pageProps} />
    </div>
  );
}

MyApp.getInitialProps = async ({Component, ctx}) => {
  let pageProps = {};

  pageProps.serverSide = !!ctx.req;

  return pageProps;
};

export default MyApp;

#pages/index.js
function Index(props) {
  return (
    <div></div>
  );
}

export async function getServerSideProps(ctx) {
  console.log("Index.getServerSideProps()", !!ctx.req);

  return {props: {somePageProp: "somePagePropContent"}};
}

export default Index;
```

This causes me to believe that getServerSideProps simply takes the `props` property of it's return result, and spreads that onto `pageProps` in _app. What's amazing here is that `MyApp.getInitialProps()` does not reference the child page's `[Page].getInitialProps()`, somehow it just knows. Previously, we would manually call `await Component.getInitialProps(ctx)` inside the `MyApp.getInitialProps()`. I found this to be a nice surprise, as it's less code to write in _app.

**Comment by kristojorg** [comment]
I am also wondering this. My `_app` component must compile some global data, like creating a redux store for example, which I would then like to pass to `getServerSideProps` of the page being rendered. Not sure how to accomplish that

**Reply by willybeans** [reply]
Im also looking to do this, is it not possible?

**Reply by kristojorg** [reply]
Nope. Can't pass from `getInitialProps` to `getServerSideProps` in the pages

**Reply by willybeans** [reply]
Thanks for the prompt reply. Do you know if its possible to pass data from the _app file down to the component's getServerSideProps? I am able to do this with getInitialProps, since it has visibility into the components getInitialProps function, but this functionality does not seem to exist for getServerSideProps. Sorry to bother with a follow up, just struggling to find anything about this on google. I assume this is strictly impossible as of now since _app does not have getServerSideProps. Do you think its a fair assumption?

**Reply by kristojorg** [reply]
I don't think any communication between `getServerSideProps` and `getInitialProps` is possible at this point. I think the best you can do is switch entirely to `getServerSideProps` and either load global data (repetitively) on every page, or else stick with `getInitialProps` for now. I've chosen to stick with `getInitialProps` until there is some way to load global data with `getServerSideProps`.

I could be wrong on this, or out of date, but this was what I decided as of beginning April

**Reply by jasonwilliams** [reply]
@kristojorg did you find a solution for this?

**Reply by kristojorg** [reply]
I didn't find a solution - the two cannot communicate. Instead at this point I have a `getServerSideProps` or `getStaticProps` that runs on every page to fetch the same data plus whatever custom data the page needs

**Reply by kristojorg** [reply]
This is what I am currently using: https://github.com/vercel/next.js/discussions/11183#discussioncomment-114597

**Reply by Wolfleader101** [reply]
anyone got an answer to this yet?
I am trying to use getServerSideProps in the app so I can use a passportJS session across the whole application.

**Comment by goldensunliu** [comment]
@majelbstoat can u update the title to `_app.getServersideProps`?

**Comment by majelbstoat** [comment]
The question isn't about _app.getServersideProps - support for which doesn't exist yet. There are other discussions about adding support for that. This is about the interaction between _app.getInitialProps and page.getServersideProps.

**Comment by kristojorg** [comment]
One thing I have noticed is that when using `_app.getInitialProps` with a page `getServerSideProps`, the `_app.getInitialProps` then begins to get called whenever the client requests the server side props, so it ends up needing to deal with requests like `/_next/data/development/yourpage.json`.

Not sure if this is intended behavior or not, but it seems that `_app.getInitialProps` shouldn't be run for these requests

**Reply by fabb** [reply]
sounds like a bug, was this already fixed in newer next.js versions?

**Comment by KyruCabading** [comment]
I'm coming across this issue too.. And have been looking all over to find a way to have a custom app that calls app.getInitialProps & a page with getServerSide props.

So to confirm, this is not possible? I have a use case that would require this.

**Comment by kristojorg** [comment]
For anyone coming across this, I have make a function `withAppProps(pageGetServerSideProps)` which wraps the page's `getServerSideProps` and provides the app props. Then on a page I simply write:
```
const getServerSideProps = withAppProps((ctx) => { ... });
```

That way I can easily fetch some default props for every page. It would be better to have it live in the `_app` but since that isn't possible right now, this workaround works

**Reply by torregianif** [reply]
Can you please provide a more detailed example of this? I am having the same issue and don't want to call getServerSideProps in each page. Thanks!

**Reply by kristojorg** [reply]
Sure, my `withAppProps` looks like this: 
```ts
export default function withAppProps(
  pageGetStaticProps?: GetStaticProps
): GetStaticProps<AppProps> {
  return async ctx => {
      // fetch the static props for the page
      const pageData = (await pageGetStaticProps?.(ctx)) ?? { props: {} };
      return {
        // revalidate library-wide data once per hour per route
        revalidate: 60 * 60,
        // overwrite with page settings
        ...pageData,
        props: {
          ...pageData.props,
          globalMessage: "Hello I'm the App"
        }
      };
  };
}
```

I can then use it like this:
```ts
// pages/home.tsx

export default function Home({globalMessage, message}) {...}

export const getStaticProps = withAppProps((ctx) => {
  return {
    props: {
      message: "hello world"
    },
  }
});

**Reply by gremo** [reply]
I came into the same problem and this is a very smart solution, thank you! 👍 I need `withAppProps` for rendering social links in the site footer, and the links came from API.

The only drawback is that you need to actually remember to define  `export const getStaticProps`  for every page. Any chance to say (maybe using `_app.js`) "if the page doesn't define `getStaticProps`, use `withAppProps` passing an empty callback"?

**Reply by DopamineDriven** [reply]
I'm not sure if that's an ideal fix since using the withAppProps might opt you out of automatic static optimization. If it doesn't then that should work. That or a HOC you append onto the export default __ might work.  Does using your withAppProps function result in opting-out of static optimization?

**Reply by gremo** [reply]
@DopamineDriven if you use `withAppProps` doesn't opt-out of static optimizations, it still generates SSG/ISR pages.

However, as I said, the main problem with this approach is that you need to actually remember to use it **in every page**, or the template could break. Another downside is that you can't have your layout defined in the `_app.js` (imagine that your server side data returns something you need to use inside a `<Footer>` component).

**Reply by kristojorg** [reply]
@gremo is right. It's frustrating to have to use on every page, and frustrating that the layout has to be wrapped around individual pages, but it's the best solution I could find to the problem.

**Reply by gremo** [reply]
I've searched a lot about this and I can confirm that no better solution exists ATM. I'm going to test with context and context providers but I fear it's not going to solve anything.

Maybe Next it's fine for a mixed static/dynamic website (I don't like much this idea), but it would be very very cool to have a full dynamic website with ISR and a relative low timeout! Otherwise, we should give up and go for a full SSR website.

**Comment by kmvan** [comment]
Is there no easy way? I don't like hard code very much.

**Comment by DopamineDriven** [comment]
This is what I have so far. Trying to get implicit cross-function type retention like I do on the client (pushing the provided apolloState into returned props for client inference)

### _app.tsx
```tsx
import '@/styles/index.css';
import '@/styles/chrome-bug.css';
import 'keen-slider/keen-slider.min.css';

import App, {
	AppContext,
	AppInitialProps,
	AppProps,
	NextWebVitalsMetric
} from 'next/app';
import { useRouter } from 'next/router';
import { ApolloProvider } from '@apollo/client';
import { useEffect, FC } from 'react';
import {
	useApollo,
	initializeApollo,
	addApolloState
} from '@/lib/apollo';
import * as gtag from '@/lib/analytics';
import { MediaContextProvider } from '@/lib/artsy-fresnel';
import { Head } from '@/components/Head';
import { GTagPageview } from '@/types/analytics';
// import { AppLayout } from '@/components/Layout';
import {
	GetInitialPropsResult,
	InferGetInitialPropsType
} from '@/types/next-abstracted';
import {
	DynamicNavQuery,
	DynamicNavQueryVariables,
	DynamicNavDocument,
	WordpressMenuNodeIdTypeEnum
} from '@/graphql/generated/graphql';

const Noop: FC = ({ children }) => <>{children}</>;

function NextApp({
	Component,
	pageProps: { ...pageProps },
}: AppProps &
	InferGetInitialPropsType<typeof NextApp.getInitialProps>) {
	const apolloClient = useApollo(pageProps);
	const LayoutNoop = (Component as any).LayoutNoop || Noop;
	const router = useRouter();
	useEffect(() => {
		document.body.classList?.remove('loading');
	}, []);
	useEffect(() => {
		const handleRouteChange = (url: GTagPageview) => {
			gtag.pageview(url);
		};
		router.events.on('routeChangeComplete', handleRouteChange);
		return () => {
			router.events.off('routeChangeComplete', handleRouteChange);
		};
	}, [router.events]);
	return (
		<>
			<Head />
			<ApolloProvider client={apolloClient}>
				<MediaContextProvider>
					<LayoutNoop pageProps={pageProps}>
						<Component {...pageProps} />
					</LayoutNoop>
				</MediaContextProvider>
			</ApolloProvider>
		</>
	);
}

NextApp.getInitialProps = async (
	appContext: AppContext
): Promise<
	GetInitialPropsResult<{
		pageProps: AppInitialProps;
		Header: DynamicNavQuery['Header'];
		Footer: DynamicNavQuery['Footer'];
	}>
> => {
	const pageProps = await App.getInitialProps(appContext);

	const graphqlClient = initializeApollo();
	const dynamicNav = await graphqlClient.query<
		DynamicNavQuery,
		DynamicNavQueryVariables
	>({
		query: DynamicNavDocument,
		variables: {
			idHead: 'Header',
			idTypeHead: WordpressMenuNodeIdTypeEnum.NAME,
			idTypeFoot: WordpressMenuNodeIdTypeEnum.NAME,
			idFoot: 'Footer'
		}
	});
	return addApolloState(graphqlClient, {
		pageProps: {
			...pageProps.pageProps,
			Header: dynamicNav.data.Header,
			Footer: dynamicNav.data.Footer
		}
	});
};

export function reportWebVitals(metric: NextWebVitalsMetric) {
	console.debug('vital: ', metric);
}

export default NextApp;
```

### index.tsx (for comparison, this does provide 100% cross-function types on InferGet*method*PropsType)
```tsx
import { Container, LandingHero } from '@/components/UI';
import { initializeApollo, addApolloState } from '@/lib/apollo';
import { ServicesTopCoalesced } from '@/components/ServicesTop';
import { AppLayout } from '@/components/Layout';
import {
	GetStaticPropsContext,
	GetStaticPropsResult,
	InferGetStaticPropsType
} from 'next';
import {
	ServicesTopQuery,
	ServicesTopQueryVariables,
	ServicesTopDocument,
	DynamicNavDocument,
	DynamicNavQueryVariables,
	DynamicNavQuery,
	WordpressMenuNodeIdTypeEnum
} from '@/graphql/generated/graphql';

export function Index({
	other,
	popular,
	Header,
	Footer
}: InferGetStaticPropsType<typeof getStaticProps>) {
	return (
		<>
			<AppLayout
				title={'✂ The Fade Room Inc. ✂'}
				Header={Header}
				Footer={Footer}
				hero={<LandingHero />}
			>
				<Container clean className='fit'>
					<ServicesTopCoalesced other={other} popular={popular} />
				</Container>
			</AppLayout>
		</>
	);
}

export async function getStaticProps(
	ctx: GetStaticPropsContext
): Promise<
	GetStaticPropsResult<{
		other: ServicesTopQuery['other'];
		popular: ServicesTopQuery['popular'];
		Header: DynamicNavQuery['Header'];
		Footer: DynamicNavQuery['Footer'];
	}>
> {
	const apolloClient = initializeApollo();
	await apolloClient.query<
		DynamicNavQuery,
		DynamicNavQueryVariables
	>({
		query: DynamicNavDocument,
		variables: {
			idHead: 'Header',
			idTypeHead: WordpressMenuNodeIdTypeEnum.NAME,
			idTypeFoot: WordpressMenuNodeIdTypeEnum.NAME,
			idFoot: 'Footer'
		}
	});

	await apolloClient.query<
		ServicesTopQuery,
		ServicesTopQueryVariables
	>({
		query: ServicesTopDocument,
		variables: {
			other: 'all',
			popular: 'popular-service'
		}
	});
	return addApolloState(apolloClient, {
		props: {},
		revalidate: 10
	});
}
export default Index;
```

I'd like to move  AppLayout into App with 100% strongly typed props passed from NextApp.getInitialProps to the default NextApp function being exported. 

Is there a way to explicitly state the Return props expected as a promise like you can with GetStaticPropsResult and GetServerSidePropsResult?

Update:

Now my Component is typed as follows: 
```tsx
var Component: NextComponentType<NextPageContext, any, (appContext: AppContext) => Promise<GetInitialPropsResult<{
    pageProps: AppInitialProps;
    Header: Maybe<{
        __typename?: "WordpressRootQuery" | undefined;
    } & {
        ...;
    }>;
    Footer: Maybe<...>;
}>>>
```

Note: I am trying to write my own GetInitialPropsResult and InferGetInitialPropsType definitions...kind of works at the moment but feedback would be _greatly_ appreciated 
### types/next-abstracted.ts
```ts
import App, {
	AppInitialProps,
	AppProps,
	AppContext
} from 'next/app';
import { Component } from 'react';

export declare function getNextAppInitialProps({
	Component,
	ctx
}: AppContext): Promise<AppInitialProps>;

export type GetInitialPropsResult<P> = {
	pageProps: P | AppInitialProps;
};

export type InferGetInitialPropsType<T> = T extends App<
	infer P,
	AppInitialProps,
	any
>
	? P
	: T extends (
			context?: AppContext
	  ) => Promise<GetInitialPropsResult<infer P | AppProps<T>>>
	? P
	: Component;
```

**Comment by hernandemonteiro** [comment]
## send props to _app with `getInitialProps`...
i needed run two functions and send the result to `_app`, after
searchs, i find this method:
```
MyApp.getInitialProps = ({ Component, ctx }) => {
  const dirs = readDirs("./src/pages/docs/");
  const navDir = readNavDir("./src/pages/docs");
  let pageProps = { dirs: dirs, navDir: navDir };

  return pageProps;
};
```
> in this case `MyApp` is your principal function in _app file;

## All _app code:

```
function MyApp(props) {
  const route = useRouter();
  return (
    <>
      {route.asPath.match("/docs") ? (
        <TemplateDocs
          dirs={props.navDir}
          subDirs={props.dirs}
          backgroundColorDefined="black"
          projectName={"Hernande Monteiro - Personal Blog"}
        >
          <props.Component {...props.pageProps} />
        </TemplateDocs>
      ) : (
        <props.Component {...props.pageProps} />
      )}
    </>
  );
}

MyApp.getInitialProps = ({ Component, ctx }) => {
  const dirs = readDirs("./src/pages/docs/");
  const navDir = readNavDir("./src/pages/docs");
  let pageProps = { dirs: dirs, navDir: navDir };

  return pageProps;
};

export default MyApp;

```
